cli/internal/kubernetes/upgrade.go

```diff
@@ -85,10 +86,22 @@ type Upgrader struct {
 	outWriter        io.Writer
 	tfUpgrader       *upgrade.TerraformUpgrader
 	log              debugLog
+	upgradeID        string
 }
 
 // NewUpgrader returns a new Upgrader.
-func NewUpgrader(ctx context.Context, outWriter io.Writer, log debugLog) (*Upgrader, error) {
+// If checkOnly is true, the upgrader will only create the resources necessary for upgrade checking. (i.e. no Terraform client)
+// Not creating the Terraform client in the check-case makes sense as the initialization of the Terraform client copies the
+// executable to the upgrade directory, which is not necessary for the check-only case.
+func NewUpgrader(ctx context.Context, outWriter io.Writer, log debugLog, checkOnly bool) (*Upgrader, error) {
```

Comment:
I agree that we might not always want to initialize the terraform client here. But if we don't, we should guard all access to `tfUpgrade` with a null-pointer check. `if u.tfUpgrade == nil {...}`. Alternatively you could omit the flag, put the initialization into a function and call that init function in case `u.tfUpgrade` is nil.

One could also argue that our `upgrade check` is incomplete, as it does not check if there are migrations that would need to be applied. If I am not mistaken we are applying migrations during `upgrade apply` unconditionally. Meaning that it makes no difference if the user is only upgrading services, image or k8s - migrations are applied in any case. Because of that, `upgrade check` could do a dry-run `terraform apply` to inform the user about the upcoming migration. I don't expect that you do this right now. I just had this thought and would like to hear what others think.

Reply:
re. `upgrade check`:
We could certainly do this! However I think we should not add the "cloud resources" as a singular upgradable component such as microservices, k8s, etc. - Since the CLI gets upgraded either way, it might expect the cloud resources to exist no matter which component got upgraded. I think both `upgrade check` and `upgrade apply` should always consider the Terraform migrations.

Reply:
re. nil check:
The main reason for not initializing the Terraform client here was that in the case of an `upgrade check`, the Terraform upgrade directories were created as well, as they get initialized with the Terraform client. If we dry-run the migrations in `upgrade check`, this would be obsolete.

Reply:
As it probably needs some more code changes which are out of scope for this PR, I track this in [AB#3226](https://dev.azure.com/Edgeless/Edgeless/_workitems/edit/3226).